File: src/main/java/com/vectorcast/plugins/vectorcastexecution/common/VcastUtils.java
/*
 * The MIT License
 *
 * Copyright 2020 Vector Software, East Greenwich, Rhode Island USA
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package com.vectorcast.plugins.vectorcastexecution.common;

import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.net.URISyntaxException;
import java.util.Optional;
import java.util.jar.JarFile;
import java.util.logging.Level;
import java.util.logging.Logger;
import hudson.PluginWrapper;
import hudson.security.Permission;
import hudson.security.PermissionGroup;
import hudson.security.PermissionScope;
import jenkins.model.Jenkins;

import hudson.model.Label;
import hudson.model.AutoCompletionCandidates;

import com.vectorcast.plugins.vectorcastexecution.Messages;

/** Utility class for VectorCAST. */
public class VcastUtils {

    /** Logger for plugin diagnostics. */
    private static final Logger LOGGER = Logger.getLogger(VcastUtils.class
        .getName());

    /** Plugin short name used by Jenkins' plugin manager. */
    private static final String PLUGIN_SHORT_NAME = "vectorcast-execution";

    /** Jenkins version where the monochrome icon became appropriate. */
    private static final int MONOCHROME_ICON_MAJOR = 2;

    /** Jenkins version where the monochrome icon became appropriate. */
    private static final int MONOCHROME_ICON_MINOR = 361;

    /** Permission of current view. */
    private static volatile Permission viewPermission;

   /**
     * Get the current view permissions.
     * @return Permission for current view
     */
    public static synchronized Permission getViewPermission() {
        if (viewPermission == null) {
            PermissionGroup group = new PermissionGroup(
                VcastUtils.class,
                Messages._VectorCASTRootAction_PermissionGroup()
            );

            viewPermission = new Permission(
                group,
                "View",
                Messages._VectorCASTRootAction_ViewPermissionDescription(),
                Jenkins.ADMINISTER,
                true,
                new PermissionScope[]{PermissionScope.JENKINS}
            );
        }
        return viewPermission;
    }

    /**
     * Gets the version of the plugins.
     * @return Optional returns the version
     */
    public static Optional<String> getVersion() {
        Jenkins instance = Jenkins.getInstanceOrNull();
        if (instance != null) {
            PluginWrapper plugin = instance.getPluginManager().getPlugin(
                PLUGIN_SHORT_NAME);
            if (plugin != null) {
                return Optional.of(plugin.getVersion());
            }
        }

        String implementationVersion = VcastUtils.class.getPackage()
            .getImplementationVersion();
        if (implementationVersion != null) {
            return Optional.of(implementationVersion);
        }

        URL source = VcastUtils.class.getProtectionDomain().getCodeSource()
            .getLocation();
        if (!"file".equals(source.getProtocol())
                || !source.getPath().toLowerCase().endsWith(".jar")) {
            return Optional.empty();
        }
        try (JarFile archive = new JarFile(new File(source.toURI()))) {
            return Optional.ofNullable(archive.getManifest()
                .getMainAttributes().getValue("Plugin-Version"));
        } catch (IOException | URISyntaxException ex) {
            LOGGER.log(Level.FINE, "Unable to read VectorCAST plugin version",
                ex);
            return Optional.empty();
        }
    }

    /**
     * Returns the icon appropriate for the current Jenkins UI and user.
     *
     * @return icon resource path, or {@code null} when the action is hidden
     */
    public static String getVectorCASTIconFileName() {
        if (!Jenkins.get().hasPermission(getViewPermission())) {
            return null;
        }
        return supportsMonochromeIcons(Jenkins.VERSION)
            ? "/plugin/vectorcast-execution/icons/vector_favicon_bw.png"
            : "/plugin/vectorcast-execution/icons/vector_favicon.png";
    }

    private static boolean supportsMonochromeIcons(final String version) {
        String[] components = version.split("\\.");
        try {
            int major = Integer.parseInt(components[0]);
            int minor = Integer.parseInt(components[1]);
            return major > MONOCHROME_ICON_MAJOR
                || (major == MONOCHROME_ICON_MAJOR
                && minor >= MONOCHROME_ICON_MINOR);
        } catch (NumberFormatException | ArrayIndexOutOfBoundsException ex) {
            LOGGER.log(Level.FINE, "Unable to parse Jenkins version: {0}",
                version);
            return false;
        }
    }

    /**
     * Default constructor for subclasses.
     * @throws UnsupportedOperationException is called
     */
    protected VcastUtils() {
        // prevents calls from subclass
        throw new UnsupportedOperationException();
    }

    /**
     * Update the potential labels to be used.
     * @param value @QueryParameter String
     * @return AutoCompletionCandidates with the list of the potential
     *         node matches
     */
    public static AutoCompletionCandidates completeNodeLabel(
            final String value) {

        AutoCompletionCandidates c = new AutoCompletionCandidates();

        for (Label l : Jenkins.get().getLabels()) {
            if (l.getName().startsWith(value)) {               
                c.add(l.getName());
            }
        }
        return c;
    }

}

File: src/main/resources/scripts/vcast_utils.py
#
# The MIT License
#
# Copyright 2026 Vector Informatik, GmbH.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

import os
import inspect

def __get_script_filename():
    """Return 'filename::function#line' of the caller. Compatible with Python 2.7-3.9."""
    try:
        stack = inspect.stack()
        # [2] = caller of the function that called this
        frameinfo = stack[2]

        # Handle both 2.x and 3.x attribute names
        filename = getattr(frameinfo, 'filename', frameinfo[1])
        funcname = getattr(frameinfo, 'function', frameinfo[3])
        lineno = getattr(frameinfo, 'lineno', frameinfo[2])
    except Exception:
        filename = '<unknown>'
        funcname = '<unknown>'
        lineno = -1
    finally:
        # Clean up to prevent reference cycles (important for Py2)
        del stack

    return "%s::%s#%s" % (os.path.basename(filename), funcname, lineno)

def checkProjectResults(vcproj):
    
    anyLocalResults = False
    anyImportedResults = False

    try:
        results = vcproj.project.repository.get_full_status([])
        all_envs = []
        for env in vcproj.Environment.all():
            if env.is_active:
                all_envs.append(env.level._full_path)

        for result in results:
            if result in all_envs:
                if results[result]['local'] != {}:
                    anyLocalResults = True

                if results[result]['imported'] != {}:
                    anyImportedResults = True
    except Exception as e:
        print(e)
        
    return anyLocalResults, anyImportedResults
    
def checkVectorCASTVersion(minimumVersion, quiet = False):
    
    encFmt = getVectorCASTEncoding()
    tool_version = os.path.join(os.environ['VECTORCAST_DIR'], "DATA", "tool_version.txt")
    with open(tool_version,"rb") as fd:
        ver = fd.read().decode(encFmt,"replace")
    
    try:
        verNo = int(ver.split(" ",1)[0])
    except:
        verNo = int(ver.split(".",1)[0])

    if verNo >= minimumVersion:
        if not quiet:
            print("Running with VC Version: " + ver);
        return True
        
    if not quiet:
        print("\nXXX " + __get_script_filename() + "    requires VectorCAST Version 2021 or higher.\nXXX\nXXX   Using version: " + ver)
    
    return False

def dump(obj):
    if hasattr(obj, '__dict__'): 
        return vars(obj) 
    else:
        try:
            return {attr: getattr(obj, attr, None) for attr in obj.__slots__} 
        except:
            return str(obj)

def fmt_percent(num, dom):
    pct = 0.0
    if (dom > 0.0):
        pct = 100.0 * num / dom
        pct_round = round(pct, 2)
        str_num = str(pct_round)
    else:
        str_num = "0%"
    return str_num
    
def getVectorCASTEncoding():

    import locale
    
    # get the VC langaguge and encoding
    enc = locale.getpreferredencoding(False)
        
    return enc;
    
def printVectorLogo():
    print( "                                                                                  ####                             ")
    print( "                                                                                  ########                         ")
    print( "                                                                                  ###########                      ")
    print( "          @@@    @@@   @@@@@@@@    @@@@@@@ @@@@@@@@@@  @@@@@@@@    @@@@@@@@          ###########                   ")
    print( "           @@@@ @@@    @@@        @@@         @@@     @@@@  @@@@   @@@   @@@            ###########                ")
    print( "            @@@@@@     @@@@@@    @@@          @@@    @@@     @@@   @@@@@@@@@               ##########              ")
    print( "             @@@@      @@@        @@@         @@@     @@@@  @@@@   @@@@@@@@             ###########                ")
    print( "              @@       @@@@@@@@    @@@@@@@    @@@      @@@@@@@@    @@@  @@@@         ###########                   ")
    print( "                                                                                  ###########                      ")
    print( "                                                                                  ########                         ")
    print( "                                                                                  ####                             ")

if __name__ == '__main__':

    checkVectorCASTVersion(21)